src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -0,0 +1,352 @@
+                var rootNamespaceSymbol = assemblySymbol.GlobalNamespace;
+
+                if (key == null)
+                {
+                    // Can't cache items for reference with null key, so just create them and return. 
+                    var items = GetCompletionItemsForTopLevelTypeDeclarations(rootNamespaceSymbol, cancellationToken);
+                    HandleAccessibleItems(items, isInternalsVisible, handleAccessibleItem);
+                }
+
+                var checksum = SymbolTreeInfo.GetMetadataChecksum(solution, peReference, cancellationToken);
+                GetAccessibleTopLevelTypesWorker(
+                    key,
+                    rootNamespaceSymbol,
+                    checksum,
+                    isInternalsVisible,
+                    handleAccessibleItem,
+                    _peItemsCache,
+                    cancellationToken);
+
+                static string GetReferenceKey(PortableExecutableReference reference)
+                    => reference.FilePath ?? reference.Display;
+            }
+
+            /// <summary>
+            /// Filter out inaccessible items and pass the rest to the callback
+            /// </summary>
+            private static void HandleAccessibleItems(
+                ImmutableArray<CompletionItemEntry> items,
+                bool isInternalsVisible,
+                Action<CompletionItem> handleAccessibleItem)
+            {
+                for (var i = 0; i < items.Length; ++i)
+                {
+                    var item = items[i];
+                    if (item.IsPublic || isInternalsVisible)
+                    {
+                        handleAccessibleItem(item.Item);
+                    }
+                }
+            }
+
+            private static void GetAccessibleTopLevelTypesWorker<TKey>(
+                TKey key,
+                INamespaceSymbol rootNamespace,
+                Checksum checksum,
+                bool isInternalsVisible,
+                Action<CompletionItem> handleAccessibleItem,
+                ConcurrentDictionary<TKey, ReferenceCacheEntry> cache,
+                CancellationToken cancellationToken)
+            {
+                // Cache miss, create all requested items.
+                if (!cache.TryGetValue(key, out var cacheEntry) ||
+                    cacheEntry.Checksum != checksum)
+                {
+                    var items = GetCompletionItemsForTopLevelTypeDeclarations(rootNamespace, cancellationToken);
+                    cacheEntry = new ReferenceCacheEntry(checksum, items);
+                    cache[key] = cacheEntry;
+                }
+
+                HandleAccessibleItems(cacheEntry.CachedItems, isInternalsVisible, handleAccessibleItem);
+            }
+
+            private static ImmutableArray<CompletionItemEntry> GetCompletionItemsForTopLevelTypeDeclarations(
+                INamespaceSymbol rootNamespaceSymbol,
+                CancellationToken cancellationToken)
```

Comment:
do you need the CompletionItemEntry?  Why not just make CompletionItems, and then just add them to the context on demand when requested?

Reply:
n/m.  i see that the entry is to also store if ti's public or not.